ssh/agent.go

```diff
@@ -0,0 +1,208 @@
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/vladimirvivien/echo"
+)
+
+// ssh-agent constant identifiers
+const (
+	AgentPidIdentifier = "SSH_AGENT_PID"
+	AuthSockIdentifier = "SSH_AUTH_SOCK"
+)
+
+var (
+	// singleton instance of agent
+	current *agent
+	// mutex to update the global agent variable
+	agentMux sync.Mutex
+)
+
+// setCurrentAgent sets the global variable
+func setCurrentAgent(a *agent) {
+	agentMux.Lock()
+	current = a
+	agentMux.Unlock()
+}
+
+// unsetCurrentAgent resets the global variable
+func unsetCurrentAgent() {
+	agentMux.Lock()
+	current = nil
+	agentMux.Unlock()
+}
+
+// agentInfo captures the connection information of the ssh-agent
+type agentInfo map[string]string
+
+// agent represents an instance of the ssh-agent
+type agent struct {
+	// pid of the ssh-agent
+	pid string
+
+	// Authentication socket to communicate with the ssh-agent
+	authSockPath string
+
+	// File paths of the keys added to the ssh-agent
+	keyPaths []string
+}
+
+// AddKey adds a key to the ssh-agent process
+func (agent *agent) AddKey(keyPath string) error {
+	e := echo.New()
+	sshAddCmd := e.Prog.Avail("ssh-add")
+	if len(sshAddCmd) == 0 {
+		return errors.New("ssh-add not found")
+	}
+
+	p := e.Env(agent.GetEnvVariables()).
+		RunProc(fmt.Sprintf("%s %s", sshAddCmd, keyPath))
+	if err := p.Err(); err != nil {
+		return errors.Wrapf(err, "could not add key %s to ssh-agent", keyPath)
+	}
+	agent.keyPaths = append(agent.keyPaths, keyPath)
+	return nil
+}
+
+// RemoveKey removes a key from the ssh-agent process
+func (agent *agent) RemoveKey(keyPath string) error {
+	e := echo.New()
+	sshAddCmd := e.Prog.Avail("ssh-add")
+	if len(sshAddCmd) == 0 {
+		return errors.New("ssh-add not found")
+	}
+
+	p := e.Env(agent.GetEnvVariables()).
+		RunProc(fmt.Sprintf("%s -d %s", sshAddCmd, keyPath))
+	if err := p.Err(); err != nil {
+		return errors.Wrapf(err, "could not add key %s to ssh-agent", keyPath)
+	}
+
+	return nil
+}
+
+// Stop kills the ssh-agent process.
+// It also tries to remove the added keys from the agent
+func (agent *agent) Stop() error {
+	// unset the global variable
+	defer unsetCurrentAgent()
+
+	for _, path := range agent.keyPaths {
+		logrus.Debugf("removing key from ssh-agent: %s", path)
+		err := agent.RemoveKey(path)
+		if err != nil {
+			logrus.Warnf("failed to remove SSH key from agent: %s", err)
+		}
+	}
+
+	logrus.Debugf("stopping the ssh-agent with pid: %s", agent.pid)
+	p := echo.New().Env(agent.GetEnvVariables()).RunProc("ssh-agent -k")
+
+	return p.Err()
+}
+
+// GetEnvVariables returns the space separated key=value information used to communicate with the ssh-agent
+func (agent *agent) GetEnvVariables() string {
+	return fmt.Sprintf("%s=%s %s=%s", AgentPidIdentifier, agent.pid, AuthSockIdentifier, agent.authSockPath)
+}
+
+// GetAgent returns the current singleton instance of the ssh-agent
+// if the global instance of the agent is uninitialized, then it
```

Comment:
nit: instead of `global` let's call it a `package-scoped`, global is a crashd-wide.